tools/shared.py

```diff
@@ -2652,21 +2652,44 @@ def safe_copy(src, dst):
   if dst == '/dev/null': return
   shutil.copyfile(src, dst)
 
-def clang_preprocess(filename):
-  # TODO: REMOVE HACK AND PASS PREPROCESSOR FLAGS TO CLANG.
-  return subprocess.check_output([CLANG_CC, '-DFETCH_DEBUG=1', '-E', '-P', '-C', '-x', 'c', filename])
-
-def read_and_preprocess(filename):
-  f = open(filename, 'r').read()
-  pos = 0
-  include_pattern = re.compile('^#include\s*["<](.*)[">]\s?$', re.MULTILINE)
-  while(1):
-    m = include_pattern.search(f, pos)
-    if not m:
-      return f
-    included_file = open(os.path.join(os.path.dirname(filename), m.groups(0)[0]), 'r').read()
+def find_lines_with_preprocessor_directives(file):
+  return filter(lambda x: '#if' in x or '#elif' in x, open(file, 'r').readlines())
+
+def run_c_preprocessor_on_file(src, dst):
+  # Run LLVM's C preprocessor on the given file, expanding #includes and variables found in src/setting.js.
```

Comment:
> (1) we can do it without hacks,

I am not sure the "preprocess only src/settings.js defines that are present on lines starting with `#if` and `#elif`" warrants to be called a hack over the JS one, since that was added to implement exactly the same semantic that the existing JS one gives so that these two align.

> (2) we have one preprocessor for everything

This sounds like a good direction, although it is too much to work on right now, so let me pause here, and splice the fix to the multithreading stack cookie setting to a separate PR, and pick this up when I have the cycles to start this kind of refactoring.